### PR TITLE
refactor!: rewrite `pl$col`

### LINKED ISSUE #912: Rewrite `pl$col()`
The `name` argument may cause unexpected behavior.

### PR DESCRIPTION
**Pull request by eitsupi**
Close #912

### CODE REVIEW

**Review by etiennebacher — COMMENTED** [review]
Thanks, that's a very nice refactoring! I'm a bit worried about the performance impact since `pl$col()` is basically used all the time (and for example the refactor in #904 lead to a slowdown, see #926). Can you provide some benchmarks for before/after refactor?

The [`cross`](https://github.com/davisVaughan/cross) package is very useful to easily benchmark branches.

**Comment by eitsupi** [comment]
In 05ce62d

``` r
bench::mark(
  char = {
    polars::pl$col(rep_len(letters, 1e3)) |>
      invisible()
  },
  dtype = {
    polars::pl$col(rep_len(list(polars::pl$String), 1e3)) |>
      invisible()
  },
  check = FALSE
)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 char         1.38ms   1.66ms    548.      6.52MB     1.99
#> 2 dtype      660.97ms 660.97ms      1.51  605.22KB     6.05
```

<sup>Created on 2024-03-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

In 59457c7e4a2893cc1fd491114e2aa79614f1d46a

``` r
bench::mark(
  char = {
    polars::pl$col(rep_len(letters, 1e3)) |>
      invisible()
  },
  dtype = {
    polars::pl$col(rep_len(list(polars::pl$String), 1e3)) |>
      invisible()
  },
  check = FALSE
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 char         1.32ms   1.55ms      595.    6.43MB     2.06
#> 2 dtype        3.02ms   3.52ms      252.    7.86KB     2.03
```

<sup>Created on 2024-03-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by etiennebacher** [comment]
> The performance is worse when RPolarsDataType is given as an argument because all type checks are performed

That's annoying but specifying a DataType should select all columns that match this DataType, right? If so I think it doesn't make sense to benchmark the performance when we pass a list with 1000 `pl$String`. What happens if you do `pl$col(list(pl$Float64, pl$Int16, pl$Binary, pl$String, pl$Categorical()))` on a DataFrame with many columns?

**Comment by eitsupi** [comment]
> That's annoying but specifying a DataType should select all columns that match this DataType, right? If so I think it doesn't make sense to benchmark the performance when we pass a list with 1000 `pl$String`. What happens if you do `pl$col(list(pl$Float64, pl$Int16, pl$Binary, pl$String, pl$Categorical()))` on a DataFrame with many columns?

Given that the same Expr is returned, I think testing in a DataFrame would only add extra processing.

File: R/functions__lazy.R [hunk 1]
```diff
@@ -60,60 +60,77 @@ pl_all = function(name = NULL) {
...
+
+  dots = list2(...)
+
+  if (lapply(dots, is.character) |> Reduce(`&&`, x = _)) {
+    if (length(dots) == 1L && length(dots[[1]]) == 1L) {
+      res = create_col(dots[[1]])
+    } else {
+      res = create_cols_from_strs(unlist(dots))
+    }
+  } else if (length(dots) == 1L && is.list(dots[[1]]) && lapply(dots[[1]], is_polars_dtype) |> Reduce(`&&`, x = _)) {
```
[the remark this answers is not in the document]
Comment (eitsupi) [reply R/functions__lazy.R:116]: I don't see the need to allow list since the characters can be vectors.

[the remark this answers is not in the document]
Comment (eitsupi) [reply R/functions__lazy.R:116]: I do not understand why you want to use list instead of vector. Please provide a reasonable use case.

**Comment by eitsupi** [comment]
In 88973d411ed6aa6c716d0e4a9230c62e5ccb3d49

``` r
df <- do.call(
  polars::pl$DataFrame,
  rep_len(list(polars::as_polars_series(letters)), 1e3)
)

bench::mark(
  char = {
    df$select(polars::pl$col("^_99.$"))
  },
  dtypes = {
    df$select(
      polars::pl$col(
        polars::pl$Float64,
        polars::pl$Int16,
        polars::pl$Binary,
        polars::pl$String,
        polars::pl$Categorical()
      )
    )
  }
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 char         2.05ms   2.51ms     348.     40.1KB     2.04
#> 2 dtypes       8.99ms   9.94ms      98.7    55.9KB     4.49
```

<sup>Created on 2024-03-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

In 59457c7e4a2893cc1fd491114e2aa79614f1d46a

``` r
df <- do.call(
  polars::pl$DataFrame,
  rep_len(list(polars::as_polars_series(letters)), 1e3)
)

bench::mark(
  char = {
    df$select(polars::pl$col("^_99.$"))
  },
  dtypes = {
    df$select(
      polars::pl$col(
        polars::pl$Float64,
        polars::pl$Int16,
        polars::pl$Binary,
        polars::pl$String,
        polars::pl$Categorical()
      )
    )
  }
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 char         2.02ms   2.66ms      356.    4.23KB     4.17
#> 2 dtypes       6.02ms   6.89ms      142.    1.19KB     2.15
```

<sup>Created on 2024-03-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

File: R/functions__lazy.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (etiennebacher) [reply R/functions__lazy.R:116]: I just thought that by symmetry if one works when it's a list, then the other should too. But I don't have a strong opinion on this.

**Comment by etiennebacher** [comment]
Performance looks good but I think we can improve it a bit by removing the checks on DataTypes. Since the error will be harder to understand, we can just redirect to the message in `Err_plain()` whenever there's an error:

```r
if (lapply(dots, is.character) |> Reduce(`&&`, x = _)) {
  if (length(dots) == 1L && length(dots[[1]]) == 1L) {
    res = create_col(dots[[1]])
  } else {
    res = create_cols_from_strs(unlist(dots))
  }
} else if (length(dots) == 1L && is.list(dots[[1]])) {
  res = create_cols_from_datatypes(dots[[1]])
} else {
  res = create_cols_from_datatypes(dots)
}

if (!is_ok(res)) {
  Err_plain(
    "pl$col()'s arguments must be one of the following:\n",
    "- character vectors\n",
    "- RPolarsDataTypes\n",
    "- a list of RPolarsDataTypes"
  ) |>
    uw()
} else {
  res$ok
}
```
```r
> pl$col(pl$String, "a")
Error: Execution halted with the following contexts
   0: In R: in pl$col():
   0: During function call [pl$col(pl$String, "a")]
   1: pl$col()'s arguments must be one of the following:
      - character vectors
      - RPolarsDataTypes
      - a list of RPolarsDataTypes
```
This lowers the timing for datatype selection from 4.5ms to 3.2ms. What do you think?

File: R/functions__lazy.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (eitsupi) [reply R/functions__lazy.R:116]: My understanding is that these are not symmetrical.
Characters are a vector, but `RPolarsDataType` is not a vector and can only be enumerated by list.

**Comment by eitsupi** [comment]
In 16a1853092354ebaed384ddcf16f1c0563f915a7

``` r
bench::mark(
  char = {
    lapply(rep_len(letters, 1e3), polars::pl$col) |>
      invisible()
  },
  dtype = {
    lapply(rep_len(list(polars::pl$String), 1e3), polars::pl$col) |>
      invisible()
  },
  check = FALSE
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 char           79ms     80ms      12.5    6.48MB     25.0
#> 2 dtype        82.4ms   82.4ms      12.1   18.66KB     48.5
```

<sup>Created on 2024-03-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by eitsupi** [comment]
@etiennebacher The skipping of type checking (as in the main branch) seems to have eliminated the performance degradation.

**Review by etiennebacher — APPROVED** [review]
Cool, thanks! Feel free to merge once CI passes

**Pull request merged**